src/xdp/xsk.c

```diff
@@ -275,16 +293,60 @@ XskSignalReadyIo(
             if (IoSetCancelRoutine(Irp, NULL) == NULL) {
                 Irp = NULL;
             }
+        } else if (Xsk->NotifyCallbackArmed) {
+            OutFlags = XskWaitInFlagsToOutFlags(Xsk->IoWaitFlags & ReadyFlags);
+            NotifyCallbackCompletionContext = Xsk->NotifyCallbackCompletionContext;
+
+            Xsk->NotifyCallbackCompletionContext = NULL;
+            Xsk->NotifyCallbackArmed = FALSE;
+            Xsk->IoWaitFlags = 0;
+
+            DoCallback = TRUE;
         } else {
             (VOID)KeSetEvent(&Xsk->IoWaitEvent, IO_NETWORK_INCREMENT, FALSE);
         }
+
+        CompletedWait = TRUE;
     }
     KeReleaseSpinLock(&Xsk->Lock, OldIrql);
 
     if (Irp != NULL) {
         EventWriteXskNotifyAsyncComplete(&MICROSOFT_XDP_PROVIDER, Xsk, Irp, Irp->IoStatus.Status);
         IoCompleteRequest(Irp, IO_NETWORK_INCREMENT);
     }
+
+    if (DoCallback && !SkipCallback) {
+        Xsk->NotifyCallback(NotifyCallbackCompletionContext, OutFlags);
+    }
+
+    return CompletedWait;
+}
+
+static
```

Comment:
it might make these three routines more obviously intended for asserts if we wrap them in a `#if DBG` block

Reply:
You mean this?

```c
#if DBG

static
BOOLEAN
XskRingIsSet(
    _In_ const XSK_KERNEL_RING *Ring
    )
{
    return Ring->Size != 0 && Ring->Shared != NULL && ((Ring->UserVa != NULL) == (Ring->Mdl != NULL));
}

static
BOOLEAN
XskRingIsNotSet(
    _In_ const XSK_KERNEL_RING *Ring
    )
{
    return Ring->Size == 0 && Ring->Shared == NULL && Ring->UserVa == NULL && Ring->Mdl == NULL;
}

static
BOOLEAN
XskRingIsValid(
    _In_ const XSK_KERNEL_RING *Ring
    )
{
    return XskRingIsSet(Ring) || XskRingIsNotSet(Ring);
}

#endif
```